src/include/daos_task.h

```diff
@@ -423,74 +423,43 @@ typedef struct {
 	unsigned int		nr;
 	daos_iod_t		*iods;
 	d_sg_list_t		*sgls;
-	daos_iom_t		*maps;
-} daos_obj_fetch_t;
+	daos_iom_t		*maps; /* only valid for fetch */
+} daos_obj_rw_t;
 
-typedef struct {
-	daos_handle_t		oh;
-	daos_handle_t		th;
-	daos_key_t		*dkey;
-	unsigned int		nr;
-	daos_iod_t		*iods;
-	d_sg_list_t		*sgls;
-} daos_obj_update_t;
-
-typedef struct {
-	daos_handle_t		oh;
-	daos_handle_t		th;
-	uint32_t		*nr;
-	daos_key_desc_t		*kds;
-	d_sg_list_t		*sgl;
-	daos_anchor_t		*anchor;
-} daos_obj_list_dkey_t;
+typedef daos_obj_rw_t		daos_obj_fetch_t;
+typedef daos_obj_rw_t		daos_obj_update_t;
 
 typedef struct {
 	daos_handle_t		oh;
 	daos_handle_t		th;
 	daos_key_t		*dkey;
-	uint32_t		*nr;
+	daos_key_t		*akey;
+	uint32_t		*nr;	/* number of dkeys/akeys/kds entries */
 	daos_key_desc_t		*kds;
 	d_sg_list_t		*sgl;
-	daos_anchor_t		*anchor;
-} daos_obj_list_akey_t;
-
-typedef struct {
-	daos_handle_t		oh;
-	daos_handle_t		th;
-	daos_key_t		*dkey;
-	daos_key_t		*akey;
-	daos_size_t		*size;
-	daos_iod_type_t		type;
-	uint32_t		*nr;
-	daos_recx_t		*recxs;
-	daos_epoch_range_t	*eprs;
-	daos_anchor_t		*anchor;
-	uint32_t		*versions;
-	bool			incr_order;
-} daos_obj_list_recx_t;
-
-/* argument structure for object internal task */
-typedef struct {
-	daos_handle_t		oh;
-	daos_handle_t		th;
-	daos_key_t		*dkey;
-	daos_key_t		*akey;
 	daos_size_t		*size;	/*total buf size for sgl buf, in case
 					 *it uses bulk transfer
 					 */
-	uint32_t		*nr;	/* number of kds entries, please refer
-					 * daos_obj_list_recx() for other
-					 * parameters
-					 */
-	daos_key_desc_t		*kds;
+	daos_iod_type_t		type;
+	daos_recx_t		*recxs;
 	daos_epoch_range_t	*eprs;
-	d_sg_list_t		*sgl;
+	/* anchors for obj list -
+	 * list_dkey uses dkey_anchor,
+	 * list_akey uses akey_anchor,
+	 * list_recx uses anchor,
+	 * list_obj uses all the 3 anchors.
+	 */
 	daos_anchor_t		*anchor;
 	daos_anchor_t		*dkey_anchor;
 	daos_anchor_t		*akey_anchor;
```

Comment:
TBH, i'm not sure this change is good. I think we should not unify the task argument struct for enumeration.
this is different from update/fetch since the arguments differ more in between the 4 operation types.

Reply:
understood. the main reason why I changed that is currently the parameter passing is too complex, see dc_obj_shard_list()/dc_obj_list_internal() now both with 17 parameters. unify the parameter could make it much simpler.
I'll add more comment about the task argument struct for different enum, please check it is better for you.

Reply:
yea im aware of those functions. we are adding some API complexity in favor of better handling internally. I guess that's fine, the task API is complicated anyway so this might not be that complicated to introduce. thanks for the comments.